File: src/main/java/com/freya02/botcommands/api/core/service/InstanceSupplier.java
package com.freya02.botcommands.api.core.service;

import com.freya02.botcommands.api.BContext;
import com.freya02.botcommands.api.core.config.BServiceConfigBuilder;
import org.jetbrains.annotations.NotNull;

/**
 * Functional interface where you supply an instance of the given class type (your command)
 *
 * @param <T> Type of the class to instantiate
 *
 * @see BServiceConfigBuilder#registerInstanceSupplier(Class, InstanceSupplier)
 */
public interface InstanceSupplier<T> {
    @NotNull
    T supply(@NotNull BContext context);
}

File: src/main/kotlin/io/github/freya022/botcommands/api/core/service/InstanceSupplier.kt
package io.github.freya022.botcommands.api.core.service

import io.github.freya022.botcommands.api.core.BContext
import io.github.freya022.botcommands.api.core.config.BServiceConfigBuilder

/**
 * Supplies an instance of the registered type.
 *
 * @param T Type of the class returned
 *
 * @see BServiceConfigBuilder.registerInstanceSupplier
 */
@Deprecated("For removal, replaced by ServiceSupplier")
fun interface InstanceSupplier<T : Any> {
    fun supply(context: BContext): T
}
